### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by emaxerrno**
Hey Guys, @jcrugzz  worte the first impl for the nodejs driver. 

Lets review when we get a chance! looking fresh!

@bigs 
@RBlafford

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  //
+  // Remark: Do we want to fail if we dont get a name? For now lets generate
+  // a unique identifier
+  //
+  this.name = options.name || 'node-' + process.hrtime().join('-');
+  //
+  // Set the grouping to the correct INT code based on what we have in the
+  // thrift types. We default to SHUFFLE
+  //
+  this.grouping = groupMap(options.grouping || 'shuffle');
```
Comment (emaxerrno) [inline index.js]: I love the idea of having the 'metadata' be part of the constructor object. :+1: 

However, I would remove the `this.grouping` as it doesn't make sense to have a grouping for ALL streams

File: index.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+  if (!this.grouping) throw new Error('Invalid grouping ' + options.grouping);
```
Comment (emaxerrno) [inline index.js]: not needed after the comment above.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+  //
+  // Computation init function and process functions which are required
+  //
+  this._init = typeof options.init === 'function' ? options.init : undefined;
+  this._processRecord = typeof options.processRecord === 'function'
+    ? options.processRecord
```
Comment (emaxerrno) [inline index.js]: we should make a comment that we validate the object later for the functions required by the driver.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    : undefined;
+
+  this.processTimer = typeof options.processTimer === 'function'
+    ? options.processTimer
+    : undefined;
+
+  if (!this._init || !this._processRecord) {
```
Comment (emaxerrno) [inline index.js]: We have this in the validation method below. No need to double check?

File: index.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  //
+  // TODO: Validate these arrays for the case that they are objects as they
+  // require both a name and grouping property
+  //
+  this.publish = options.publish || [];
+  this.subscribe = options.subscribe || [];
+
+  if (!this.publish.length && !this.subscribe.length) {
```
Comment (emaxerrno) [inline index.js:55]: w00T, this is fresh! :+1:

File: index.js [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+//
+// Inherit from an event emitter
+//
+util.inherits(Computation, EE);
+
+function Computation(options) {
+  EE.call(this);
+  //
+  // Remark: Do we want to fail if we dont get a name? For now lets generate
```
Comment (emaxerrno) [inline index.js]: yeah, for sure.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+  // a unique identifier
+  //
+  this.name = options.name || 'node-' + process.hrtime().join('-');
```
Comment (emaxerrno) [inline index.js]: we should just validate it :)

File: index.js [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+// Parse environment variables and create a thrift service based on the
+// computation
+//
+exports.run = function run(computation) {
+  var server = thift.createServer(ComputationService, computation);
+
+  //
+  // Send a start event when we are fully connected to everything
+  //
+  var finish = after(2, function () {
```
Comment (bigs) [inline index.js:273]: what's happening here? is this just a "yield"?

Comment (jcrugzz) [inline index.js:273]: @bigs it basically waits for 2 function calls to happen before emitting. So its a simple counter behind the scenes before it executes the given callback. In this case we aren't emitting start until we have registered the metadata with the proxy and have started listening on our given port

Comment (bigs) [inline index.js:273]: ah cool!

**Comment by emaxerrno** [comment]
This looks so good. I'm going to get a chance to integrate it on friday. I wanna merge this :) 

Did the guy say anything about the concord pkg in npm?

**Comment by jcrugzz** [comment]
@senior7515 Definitely let me know how i can help with the testing. In terms of the package name, I havent gotten an email back from the developer who owns `concord`. I'd say for the near term we can go with `concord.io`, `concordio`, `concord.js` unless you have something better in mind. I can make that change or you can just modify the `package.json` so you can `npm publish` the package post merge.

**Pull request merged**